### PR TITLE
fix(behavior_path_planner): fix-lane-follow-in-behavior-path-planner

### PR DESCRIPTION
**Pull request by shulanbushangshu**
Signed-off-by: jack.song <jack.song@autocore.ai>

## Description
When the global route includes the changing lane,the function of "checkLaneIsInIntersection" will return true(because the distance is not enough by calculation) by the current code in some positions as the target and the path by "lane_following" will be wrong. But the distance of the lanes is actual enough.
The simulation is shown as below:

image

image

## Related links
https://github.com/autowarefoundation/autoware.universe/issues/2310

## Tests performed
By changing the judgment of the "checkLaneIsInIntersection",the simulation is shown as below:
case1:

lane_fowwing

case2:

lane_follow_2

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/utilities.cpp [hunk 1]
```diff
@@ -1829,6 +1829,11 @@ bool checkLaneIsInIntersection(
     if (get_signed_distance > min_lane_change_distance) {
       return false;
     }
+
+    auto back_lanelet_length = lanelet::utils::getLaneletLength2d(lanelet_sequence.back());
```
Comment (zulfaqar-azmi-t4) [inline planning/behavior_path_planner/src/utilities.cpp:1833]: Should be `const` here.

**Comment by zulfaqar-azmi-t4** [comment]
@shulanbushangshu 

Thank you very much for your contribution!

I think the solution should be fine, however, at the moment, I am not sure why the result differs depending on the location of the goal. I've check #2310 and the discussion seems to be on going.

Based on your current analysis, may i ask your help in explaining the reason?

**Comment by shulanbushangshu** [comment]
@zulfaqar-azmi-t4 
I show the process :
1.When I set the target position near the start of the lane and the start position is near the intersection,the goal route by mission_planner includes has the change lane in my map.
2.In the lane_following_module, the checkLaneIsInIntersection() is used in the  getReferencePath().
By these conditions, the checkLaneIsInIntersection() will get the true as the result(by current codes).
3.And checkLaneIsInIntersection() will caculate the "additional_length_to_add"(it is a big value by current codes) and the length will be used in the calculation of getCenterLinePath().
4.In the getCenterLinePath(), the isDeadEndLanelet() is used and the result is true because of the change lane.
The s_forward will be calculated by "std::min(s_forward, lane_length - lane_change_buffer);".Because of the  lane_change_buffer(is the additional_length_to_add),the  s_forward will small than the s_backward.So the route_handler.getCenterLinePath() will calculate  the wrong path.

So I think that the checkLaneIsInIntersection() has the problem in judging the route with change lane by just using the goal position and last lane.





intersection_sforword

intersection_length_2

**Comment by liuXinGangChina** [comment]
Hi,Mitsudome-san. @mitsudome-r 
Please help us to accelarate the viewing process.

Thanks,
Lucas

**Comment by xmfcx** [comment]
@zulfaqar-azmi-t4 could you review again?

**Comment by mitsudome-r** [comment]
@shulanbushangshu Is it possible to share the map so that I can also check the issue on my machine?

**Review by zulfaqar-azmi-t4 — APPROVED** [review]
Sorry I have overlooked this.
LGTM

**Comment by zulfaqar-azmi-t4** [comment]
@shulanbushangshu 
I shall close this PR because the function is no longer used in the current version.

**Pull request closed without merging**